### PR TITLE
fix: cache http client for ggv2 data client properly

### PR DESCRIPTION
**Pull request by junfuchen99**
**Issue #, if available:**

**Description of changes:**
- always keep a cache of http client for ggv2 data client;
- reconfigure http client only when private key or cert or root ca path changes.

**Why is this change necessary:**
We made a recent change in https://github.com/aws-greengrass/aws-greengrass-nucleus/pull/1368 to cache http client for ggv2 data client. Before it, we never closed v2 data client properly. 

However, it exposed a new issue - if other plugins are still using the data client while we are trying to close the client, then their api calls would fail with `java.lang.IllegalStateException: Connection pool shut down`. 

Specifically, when we fetch data client for the first time, we update data plane port with the default port number (if not previously configured), which is a device config change and triggers a client cleanup and causes api call failure.
`thing group helper calls fetchGreengrassV2DataClient -> device config change triggers clean client -> thing group helper fails to use the client and call listThingGroupsForCoreDevice`

**How was this change tested:**
Failing E2E tests are now passing.

- [ ] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [ ] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/util/GreengrassServiceClientFactory.java [hunk 1]
```diff
@@ -184,5 +204,6 @@ private void configureClient(DeviceConfiguration deviceConfiguration) {
             }
         }
         this.greengrassV2DataClient = clientBuilder.build();
+        this.greengrassV2DataClient.close();
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/GreengrassServiceClientFactory.java:207]: why are you building the client and instantly closing it?

Comment (junfuchen99) [reply src/main/java/com/aws/greengrass/util/GreengrassServiceClientFactory.java:207]: typo - forgot to remove this line when doing some other testing

**Review by MikeDombo — APPROVED** [review]

**Comment by shaguptashaikh** [comment]
Can you link the E2E test that caught this if you have time? I'm just worried that we're accidentally catching these problems in OTF UATs and E2E tests and want us to spend a minute on checking that we have good coverage around reconfiguring, using across plugins etc

**Comment by shaguptashaikh** [comment]
And, did we run the failing E2E test locally to verify this?

**Pull request merged**